### PR TITLE
attr: add ArticulationPoints and IsBiconnectedDigraph

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR adds two features: the ability to compute `ArticulationPoints` and as a consequence check if a digraph is biconnected. This uses a pretty straightforward depth first search originally by Tarjan.

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
The tests failed because I used a 2 argument lambda in the test file. Fixed it now.

File: doc/attr.xml [hunk 1]
```diff
@@ -8,6 +8,42 @@
...
+    <A>digraph</A>, if any, and, in particular, returns the empty list if
+    <A>digraph</A> is not connected. <P/>
+
+    Multiple edges and loops are ignored by this method. <P/>
+    
+    The method used in this operation has complexity <M>O(m+n)</M> where
+    <M>m</M> is the number of edges (counting multiple edges as one, and not
+    counting loops) and <M>n</M> is the number of vertices in the digraph.
+
+    See also <Ref Attr="ArticulationPoints"/>.
```
Comment (mtorpey) [inline doc/attr.xml]: This is a reference to itself.  Perhaps you meant `See also <Ref Attr="IsBiconnectedDigraph"/>.`

File: doc/attr.xml [hunk 2]
```diff
@@ -8,6 +8,42 @@
...
 ##
 
+<#GAPDoc Label="ArticulationPoints">
+<ManSection>
+  <Attr Name="ArticulationPoints" Arg="digraph"/>
+  <Returns>A list of vertices.</Returns>
+  <Description>
+
+    A connected digraph is <E>biconnected</E> if it is still connected (in the
+    sense of <Ref Prop="IsConnectedDigraph"/>) if any vertex is removed.
```
Comment (mtorpey) [inline doc/attr.xml]: Replace `if any vertex` by `when any vertex` to avoid a slightly confusing double-if.

File: doc/prop.xml [hunk 3]
```diff
@@ -8,6 +8,36 @@
 #############################################################################
 ##
 
+<#GAPDoc Label="IsBiconnectedDigraph">
+<ManSection>
+  <Prop Name="IsBiconnectedDigraph" Arg="digraph"/>
+  <Returns><K>true</K> or <K>false</K>.</Returns>
+  <Description>
+    A connected digraph is <E>biconnected</E> if it is still connected (in the
+    sense of <Ref Prop="IsConnectedDigraph"/>) if any vertex is removed.
```
Comment (mtorpey) [inline doc/prop.xml]: Perhaps `when any vertex`, as above.

File: gap/attr.gi [hunk 4]
```diff
@@ -8,6 +8,85 @@
...
+  counter     := 0;
+  visited     := BlistList([1 .. DigraphNrVertices(copy)], []);
+  num         := [];
+  low         := [];
+  parent      := [1];
+  points      := [];
+
+  v_stack     := [1];
+  w_stack     := [0];
+  depth       := 1;
```
Comment (mtorpey) [inline gap/attr.gi]: Many of these lines, like `points      := [];`, would not pass our own linting checks if they were in the Semigroups package, due to multiple spaces.  I don't mind, but would you rather Digraphs conformed to gaplint standard?

File: gap/attr.gi [hunk 5]
```diff
@@ -8,6 +8,85 @@
...
+# http://www.eecs.wsu.edu/~holder/courses/CptS223/spr08/slides/graphapps.pdf
+
+InstallMethod(ArticulationPoints, "for a digraph", [IsDigraph],
+function(digraph)
+  local copy, nbs, counter, visited, num, low, parent, points, v_stack,
+  w_stack, depth, 1st_root_nb, v, w, i;
+
+  if HasIsConnectedDigraph(digraph) and not IsConnectedDigraph(digraph) then
+    return [];
+  elif DigraphNrVertices(digraph) < 1 then
```
Comment (mtorpey) [inline gap/attr.gi]: Minor point: this `<` could be replaced by `<=` to avoid a tiny little bit more work.

File: gap/attr.gi (hunk 4 above)
Comment (james-d-mitchell) [inline gap/attr.gi]: Whitespace before `:=` does not violate anything in gaplint, so also not in the Semigroups package. Indeed clang-format will line up assignments in consecutive lines...

**Comment by james-d-mitchell** [comment]
I've made all the requested changes, except the thing about := which I don't believe violates any conventions we've used anywhere.

File: doc/attr.xml (hunk 2 above)
Comment (james-d-mitchell) [inline doc/attr.xml]: Good idea

File: doc/attr.xml (hunk 1 above)
Comment (james-d-mitchell) [inline doc/attr.xml]: Ooops

**Comment by mtorpey** [comment]
Good stuff!  Good to know whitespace around `:=` is flexible - I had thought it was like `..` or other operators.

**Pull request merged**